### PR TITLE
Move openamp demos

### PR DESCRIPTION
**Pull request by tnmysh**
Move apps directory from open-amp library repo to openamp-system-reference repo and modify cmake build infra structure as needed to build these apps.

### CODE REVIEW

File: examples/legacy_apps/system/generic/machine/zynqmp_r5/linker_remote.ld [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+   *(.gnu.linkonce.t.*)
+   *(.plt)
+   *(.gnu_warning)
+   *(.gcc_execpt_table)
+   *(.glue_7)
+   *(.glue_7t)
+   *(.vfp11_veneer)
+   *(.ARM.extab)
+   *(.gnu.linkonce.armextab.*)
+} > psu_ddr_S_AXI_BASEADDR
```
Comment (arnopo) [inline examples/legacy_apps/system/generic/machine/zynqmp_r5/linker_remote.ld:54]: seems not aligned with open-amp lib , any reason?
see: https://github.com/OpenAMP/open-amp/blob/main/apps/system/generic/machine/zynqmp_r5/linker_remote.ld#L85

File: examples/legacy_apps/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+collector_create (APP_INC_DIRS "")
+collector_create (APP_LIB_DEPS "")
+
+collector_create (APP_EXTRA_C_FLAGS "")
+collect (APP_INC_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/include")
+
+add_subdirectory (machine)
+add_subdirectory (system)
+add_subdirectory (tests)
+add_subdirectory (examples)
```
Comment (arnopo) [inline examples/legacy_apps/CMakeLists.txt:39]: could you move the update in your second commit and just have a copy past here?

File: examples/legacy_apps/cmake/collect.cmake [hunk 3]
```diff
@@ -0,0 +1,38 @@
+function (collector_create name base)
```
Comment (arnopo) [inline examples/legacy_apps/cmake/collect.cmake:1]: same for the cmake folder please move it in a separate patch

**Review by arnopo — COMMENTED** [review]
I think this is a step in the right direction to move forward. 
Just a request: please split the commit to make it simpler to compare with the original code.

i will try to rebuild and run application to test it

File: examples/legacy_apps/system/generic/machine/zynqmp_r5/linker_remote.ld (hunk 1 above)
Comment (tnmysh) [reply examples/legacy_apps/system/generic/machine/zynqmp_r5/linker_remote.ld:54]: For sometime open-amp demos are being built from downstream repo, and we never got chance to modify linker script here.

Original linker script isn't working. SO I have modified it. But, looks like for now I will keep original linker script, and send separate change fixing only linker script.

File: examples/legacy_apps/CMakeLists.txt (hunk 2 above)
Comment (tnmysh) [reply examples/legacy_apps/CMakeLists.txt:39]: Yes, I will modify next revision accordingly.

File: examples/legacy_apps/cmake/collect.cmake (hunk 3 above)
Comment (tnmysh) [reply examples/legacy_apps/cmake/collect.cmake:1]: Ack.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,35 @@
...
+```
+  $ mkdir -p build-openamp-demos
+  $ cd build-openamp-demos
+  $ cmake <openamp_system_reference/examples/legacy_apps> \
+        -DCMAKE_INCLUDE_PATH=<libmetal_built_include_dir> <openamp_built_include_dir> \
+        -DCMAKE_LIBRARY_PATH=<libmetal_built_lib_dir> <openamp_built_lib_dir> \
+	-DCMAKE_INSTALL_PREFIX=<open_amp_demos install dir>
+  $ make VERBOSE=1 install
+```
+
```
Comment (arnopo) [inline README.md:35]: I would add following in readme to explain how to "emulate" on a Linux PC


## Run application  on a Linux PC ###
It is possible to run application on a Linux PC to communicate between two Linux processes.

```
  $ cd $PROJECT_ROOT/target
  $ echo "################### run ping test #####################"
  $ LD_LIBRARY_PATH=./lib  ./bin/rpmsg-echo-static &
  $ sleep 1
  $ LD_LIBRARY_PATH=./lib  ./bin//msg-test-rpmsg-ping-static 1
  $ 
  $ echo "################### run ping test #####################"
  $ LD_LIBRARY_PATH=./lib  ./bin/rpmsg-nocopy-echo-static &
  $ sleep 1
  $ LD_LIBRARY_PATH=./lib  ./bin//rpmsg-nocopy-ping-static  1
```

**Review by arnopo — CHANGES_REQUESTED** [review]
I tested on my Linux PC using west and cmake. Few comments and suggestions

File: README.md [hunk 5]
```diff
@@ -1,2 +1,35 @@
 # openamp-system-reference
 End-to-end system reference material showcasing all the different aspects of OpenAMP, on multiple vendor platforms.
+
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md:3]: Correct, actually I can move this now too, as it shows how to build legacy apps.

File: README.md (hunk 4 above)
Comment (tnmysh) [reply README.md]: Done.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,35 @@
...
+```
+
+## Build open_amp
+```
+  $ mkdir -p build-openamp
+  $ cd build-openamp
+  $ cmake <openamp_source> -DCMAKE_INCLUDE_PATH=<libmetal_built_include_dir> \
+        -DCMAKE_LIBRARY_PATH=<libmetal_built_lib_dir> \
+	-DCMAKE_INSTALL_PREFIX=<open_amp install dir>
+  $ make VERBOSE=1 install
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md]: Done.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,35 @@
...
+# Build legacy_apps
+
+Legacy apps are moved from open_amp library repository.
+
+## Build libmetal
+```
+    $ mkdir -p build-libmetal
+    $ cd build-libmetal
+    $ cmake <libmetal_source> -DCMAKE_INSTALL_PREFIX=<libmetal install dir>
+    $ make VERBOSE=1 install
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md]: Done.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,35 @@
...
+```
+
+## Build open_amp
+```
+  $ mkdir -p build-openamp
+  $ cd build-openamp
+  $ cmake <openamp_source> -DCMAKE_INCLUDE_PATH=<libmetal_built_include_dir> \
+        -DCMAKE_LIBRARY_PATH=<libmetal_built_lib_dir> \
+	-DCMAKE_INSTALL_PREFIX=<open_amp install dir>
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md]: Done.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,35 @@
...
+
+# Build legacy_apps
+
+Legacy apps are moved from open_amp library repository.
+
+## Build libmetal
+```
+    $ mkdir -p build-libmetal
+    $ cd build-libmetal
+    $ cmake <libmetal_source> -DCMAKE_INSTALL_PREFIX=<libmetal install dir>
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md]: Done.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,35 @@
 # openamp-system-reference
 End-to-end system reference material showcasing all the different aspects of OpenAMP, on multiple vendor platforms.
+
+# Build legacy_apps
+
+Legacy apps are moved from open_amp library repository.
+
+## Build libmetal
+```
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply README.md]: Done.

File: examples/legacy_apps/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
+# Zephyr Example Application
```
Comment (arnopo) [inline examples/legacy_apps/README.md:1]: Commit message to rework it does not reflect the commit.. could be splitted in 2 commits one for the yaml, one for the doc

**Review by arnopo — COMMENTED** [review]
few new comments
after addressing them LGTM

File: examples/legacy_apps/README.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+$ make -C build VERBOSE=1 install
+```
+
+## Run application on a Linux PC
+It is possible to run application on a Linux PC to communicate between two Linux processes.
+
+```
+  $ cd $PROJECT_ROOT/target
+  $ echo "################### run ping test #####################"
+  $ LD_LIBRARY_PATH=./lib  ./bin/rpmsg-echo-static &
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/README.md:60]: Actually yes, but I am addressing in this PR by explicitly mentioning -static flag.
I think with that flag, we should be building with static libraries correct?
I will have to test this.

[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/README.md:60]: Just FYI, I haven't tested the steps in this README, I think it should be good that I test them now.

File: examples/legacy_apps/README.md (hunk 11 above)
Comment (tnmysh) [reply examples/legacy_apps/README.md:1]: Okay I agree. Probably will remove this line too.

File: examples/legacy_apps/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+Legacy apps are moved from open_amp library repository.
+
+```
+export PROJECT_ROOT=$PWD
+```
+
+## Build libmetal
+```
+  $ cd $PROJECT_ROOT/libmetal
+  $ cmake . -Bbuild -DCMAKE_INSTALL_PREFIX= $PROJECT_ROOT/target
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/README.md]: Done.

File: examples/legacy_apps/README.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+  $ make -C build VERBOSE=1
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/README.md]: Done.

File: examples/legacy_apps/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/README.md:58]: I looked into this. Looks like only passing -static isn't enough.

Some plumbing is needed in cmake infrastructure to make it work.

The issue is present on AMD platforms as well, but it doesn't affect development due to following reason:

1) tests directory isn't used in demos provided in petalinux.
2) static linkiing is happening only for remote firmware apps, and for that we have separate cmake infra that is used to build apps.

However, this issue should be fixed. I can work on that.

[the remark this answers is not in the document]
Comment (arnopo) [reply examples/legacy_apps/README.md:58]: yes issue is linked to the application build, not blocking to merge this series, can treated in a separate PR

**Review by arnopo — APPROVED** [review]

**Review by iuliana-prodan — COMMENTED** [review]
This PR looks good but why do we call the folder `legacy_apps`?
When I read legacy app, I'm thinking of something old that might not even work - but maybe is just me :)
Can we call the folder `generic_samples`, or just `generic` (since it's already in examples).


Also, not very clear for me how these samples work?
Do we have a Linux-Zephyr communication?

Can we add, for each sample a README - like we have now [here](https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/linux/rpmsg-mat-mul/README.md), for each app. 
I see the comment from the start of each sample (like [here](https://github.com/OpenAMP/openamp-system-reference/pull/44/files#diff-a1086a533ef9f1deda47d6a5e748c0558b9cdbe9c72a81ef376461cdefbcd159R6)) but IMO is not enough.

**Comment by tnmysh** [comment]
> This PR looks good but why do we call the folder `legacy_apps`? When I read legacy app, I'm thinking of something old that might not even work - but maybe is just me :) Can we call the folder `generic_samples`, or just `generic` (since it's already in examples).

Some background on this:
So, These apps used to be in open-amp repository here: https://github.com/OpenAMP/open-amp/tree/main/apps
At that time openamp-system-reference repo did not exist. Now, we have separate repo available only for demos, we are moving demos from library repository.

Ideally, we want a framework that is easily portable to any platform and any OS. This framework isn't available yet. Many demos under "legacy_apps" are hardcoded to run only on AMD-Xilinx platforms. Eventually they will be deprecated. Hence the name "legacy_apps" makes sense.

We don't want to call it "generic" for the same reason, as not all demos are easily portable. Instead, we can introduce new demos under ${REPO}/examples/{OS}/ directory that is portable on all the platforms.

> 
> Also, not very clear for me how these samples work? Do we have a Linux-Zephyr communication?
> 

Some demos are using AMD-Xilinx BSP, and so they provide Linux-Baremetal, and Linux-FreeRTOS communication. Some demos provides Linux-Linux communication.
Linux-Linux RPMSG communicaiton is tested between two linux processes. 

> Can we add, for each sample a README - like we have now [here](https://github.com/OpenAMP/openamp-system-reference/blob/main/examples/linux/rpmsg-mat-mul/README.md), for each app. I see the comment from the start of each sample (like [here](https://github.com/OpenAMP/openamp-system-reference/pull/44/files#diff-a1086a533ef9f1deda47d6a5e748c0558b9cdbe9c72a81ef376461cdefbcd159R6)) but IMO is not enough.

Yes, we should add it, but it can be part of new PR and as enhancement. This PR only moves demos as it is, and makes sure they can be compiled. So, modifying cmake build infra as needed.

File: examples/legacy_apps/cmake/modules/Findopen_amp.cmake [hunk 15]
```diff
@@ -0,0 +1,30 @@
+# Finidopen_amp
```
[the remark this answers is not in the document]
Comment (tnmysh) [reply examples/legacy_apps/cmake/modules/Findopen_amp.cmake]: Fixed.

**Comment by tnmysh** [comment]
@iuliana-prodan Thanks for reviews. Fixed comments.

**Review by iuliana-prodan — APPROVED** [review]

**Pull request merged**